flang/lib/Lower/ConvertExpr.cpp

```diff
@@ -1979,23 +2160,84 @@ class ArrayExprLowering {
                                                iters.iterVec());
     };
   }
-  CC genarr(const Fortran::evaluate::Component &) {
-    // Generate an array slice. Each step of the slice is 1 derived type
-    // size. An initial offset corresponding to the component is required. Each
-    // element of the slice array has component type.
-    TODO("component");
+
+  /// Example: <code>array%baz%qux%waldo</code>
+  CC genarr(const Fortran::evaluate::Component &x) {
+    llvm::SmallVector<mlir::Value, 8> components;
+    const auto &sym = x.GetFirstSymbol();
+    auto recTy = converter.genType(sym);
+    buildComponentsPath(components, recTy, x.base());
+    auto lf = genPathSlice(sym, components);
+    return [=](IterSpace iters) { return lf(iters); };
+  }
+  /// The `Ev::Component` structure is tailmost down to head, so the expression
+  /// <code>a%b%c</code> will be presented as <code>(component (dataref
+  /// (component (dataref (symbol 'a)) (symbol 'b))) (symbol 'c))</code>.
+  void buildComponentsPath(llvm::SmallVectorImpl<mlir::Value> &components,
+                           mlir::Type &recTy,
+                           const Fortran::evaluate::DataRef &dr) {
+    std::visit(Fortran::common::visitors{
+                   [&](const Fortran::evaluate::Component &c) {
+                     buildComponentsPath(components, recTy, c.base());
+                     auto loc = getLoc();
+                     auto name = toStringRef(c.GetLastSymbol().name());
+                     components.push_back(
+                         builder.create<fir::FieldIndexOp>(loc, name, recTy));
+                     recTy = recTy.cast<fir::RecordType>().getType(name);
```

Comment:
The propagation/mutation of the type is a bit tricky to follow, why not computing the type for the fieldIndexOp on the spot with `converter.genType(c.base().GetLastSymbol())` ? (although, maybe you would need to remove a fir.SequenceType wrapping coming from that if the base is an ArrayRef).

Reply:
It feels more natural to me to determine the type of a field using what has already been reified in the FIR code than to go back to the front-end and re-derive the type from another representation.

Reply:
OK, fine with me too.